### DISCUSSION TITLE
how to start?

### DISCUSSION THREAD

**Discussion by ghost** [vsg-dev/VulkanSceneGraph#1204] [Q&A] [answered]
Hello there!
I'm relatively a noob and I don't have big IQ, so uhm, i wanted to ask you guys (or maybe Robert himself) for some sort of a guide on how to even start with VSG. The guides out there didn't help me sadly.
My problem is that I can't compile my project.
I use VSCode on Windows, Vulkan 1.3.283, Clang 18.1.5. 
As a starting point, I tried to create a project that renders a teapot. 

So my CMakeLists contains this:
*add_subdirectory(VulkanSceneGraph)
target_include_directories(vsgtest
    PUBLIC
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}> 
        $<INSTALL_INTERFACE:include>               
    PRIVATE               
        ${CMAKE_CURRENT_SOURCE_DIR}/VulkanSceneGraph/include
)
target_link_libraries(vsgtest PUBLIC vsg::vsg)
set_target_properties(vsgtest PROPERTIES OUTPUT_NAME vsgtest)*

When i compile, there are lots of errors like 'undeclared identifier uint32\uint64', i tried adding includes like cstdlib and stdlib.h but it didn't help either, because then i keep getting more errors. What am I doing wrong? Is it even possible to do without MSVC and Visual Studio which i don't like much?

Problems come from the source files in this directory: \VulkanSceneGraph\src\glslang\glslang\MachineIndependent

Thanks!

**Comment by robertosfield** [comment]
For your first example using the [MyFirstVsgApplication](https://github.com/vsg-dev/MyFirstVsgApplication) project will be good starting place for what is required to set up the CMakeLists.txt in order to link against the VSG, vsgXchange and build your source.

**Reply by ghost** [reply]
Hello, Robert! It's a bit cringy I know, but I failed at this one, too.
I've been trying everything for a week already and failed miserably, lol.
Do i absolutely have to use Visual Studio? Because as far as I understand I need to build VSG first to link against it and I'm unable to build it for some noobish reason.

**Reply by robertosfield** [reply]
You don't need to use VisualStudio to use the VulkanSceneGraph - I'm working under Kubuntu 22.04 right now with the GNU toolchain just fine.  One of the main points of using CMake with the VSG is opens the door to cross platform development using the same build system.

Even under Windows you'll have options, but as I mostly work under Linux I'll have to defer to others for best advice for working on Windows.  This area really isn't VSG specific, it's a general C++/CMake/Windows build tools topic so you should be able to find plenty of resources online about working with Windows with/without the VisualStudio tools. 

Have you successfully compiled the VSG, or are you using a 3rd party packaging of it?

Have you been able to compile the vsgExamples and MyFirstVsgApplication?

**Reply by ghost** [reply]
Thanks,
at this point I'm just trying to compile VSG and I get 1 warning and 8 errors generated.
Like this one: VulkanSceneGraph\src\glslang\glslang\MachineIndependent/SymbolTable.h:649:18: error: unknown type name 'uint32_t'.

here are my steps to compile VSG on Windows:
PS D:\VulkanSceneGraph> cd build
PS D:\VulkanSceneGraph\build> cmake .
-- Reading 'vsg_...' macros from D:/VulkanSceneGraph/cmake/vsgMacros.cmake - look there for documentation
-- The following REQUIRED packages have been found:

 * Vulkan (required version >= 1.1.70.0)
 * Threads

-- Configuring done (0.2s)
-- Generating done (0.1s)
-- Build files have been written to: D:/VulkanSceneGraph/build
PS D:\VulkanSceneGraph\build> cmake --build .

after that i get something like glslang/glslang/MachineIndependent/glslang_tab.cpp.obj Error 1

I cloned VSG form here: https://github.com/vsg-dev/VulkanSceneGraph.git

**Comment by robertosfield** [comment]
uint32_t is standard part of C/C++ itself but there is a C version uint32_t and the C++ standard version std::uint32_t, they are exactly the same type, just latter definition adds the std namespace.

I don't know of others having the issue you are having  - I occasionally build under Windows 11 myself and haven't seen this issue with using the command line build or using VisualStudio.

What compiler version are you using?

**Reply by ghost** [reply]
you know what, I have a Linux laptop and I've just built VSG without any problems... But i can't do so on my Windows machine no matter what... I keep getting stupid mistakes, I tried everything i know to get rid of 'em, like adding includes with cstdint, nothing helps.

The only problem is that my laptop is quite old with a GTX1050m on board, while my main Windows machine has an RTX card, that's why I want to use Windows for my development. I even reinstalled Windows to see if there are any troubles with my installation.

On my Windows machine i use Clang++ 18.1.5 but I also tried latest g++ and failed, too.

if i add *cstdint* to the SymbolTable.h i get these errors:
VulkanSceneGraph\src\vsg\platform\win32\Win32_Window.cpp:202:14: error: 'VK_NAVIGATION_VIEW' was not declared in this scope  {VK_NAVIGATION_VIEW, KEY_Undefined},   // reserved

**Comment by ghost** [comment, marked as answer]
ok i was able to build it PHEWWWW
turned out I needed Windows SDK, as well as *cstdint* and *windows.h* includes added to SymbolTable.h and Win32_Window.cpp respectively.
Thanks a lot Robert for guiding me there!
Love your work.

**Comment by AnyOldName3** [comment]
Just to give some more context here, the `uint32_t` problem was caused by the old version of glslang that the VSG currently bundles, which has a problem where it used `uint32_t` without including the necessary header, as when that version was contemporary, it was transitively included by another header in all the STL implementations. libc++ and libstdc++ don't do that anymore, so now there are symptoms. This problem doesn't exist in newer versions of glslang, so when the `external_glslang` branch is merged, it will be fine.

**Reply by ghost** [reply]
thanks!
